src/main/java/duke/core/Parser.java

```diff
@@ -0,0 +1,188 @@
+     *
+     * @param userInput The String containing the user input.
+     * @param taskList  The TaskList that certain Commands will be executed on.
+     * @return The Command generated by parsing the user input.
+     * @throws InvalidCommandException if the user input is not a valid command.
+     */
+    public static Command parseUserInput(String userInput, TaskList taskList) throws InvalidCommandException {
+
+        //Check first word
+        String[] commandAndRemaining = userInput.strip().split(" ", 2);
+        String commandUpper = commandAndRemaining[0].toUpperCase();
+        Command cmd;
+        switch (commandUpper) {
+        case "ADD":
+            try {
+                cmd = parseAdd(commandAndRemaining[1].strip(), taskList);
+                break;
+            } catch (IndexOutOfBoundsException e) {
+                throw new InvalidCommandException(AddCommand.getUsage());
+            }
+        case "LIST":
+            cmd = new ListCommand(taskList);
+            break;
+        case "MARK":
+            try {
+                cmd = parseMark(commandAndRemaining[1].strip(), taskList);
+                break;
+            } catch (IndexOutOfBoundsException | NumberFormatException e) {
+                throw new InvalidCommandException(MarkCommand.getUsage());
+            }
+        case "UNMARK":
+            try {
+                cmd = parseUnmark(commandAndRemaining[1].strip(), taskList);
+                break;
+            } catch (IndexOutOfBoundsException | NumberFormatException e) {
+                throw new InvalidCommandException(UnmarkCommand.getUsage());
+            }
+        case "DELETE":
+            try {
+                cmd = parseDelete(commandAndRemaining[1].strip(), taskList);
+                break;
+            } catch (IndexOutOfBoundsException | NumberFormatException e) {
+                throw new InvalidCommandException(DeleteCommand.getUsage());
+            }
+        case "END":
+            cmd = new ExitCommand();
+            break;
+        case "FIND":
+            try {
+                cmd = new FindCommand(taskList, commandAndRemaining[1]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new InvalidCommandException(FindCommand.getUsage());
+            }
+            break;
+        default:
+            throw new InvalidCommandException();
+        }
+        return cmd;
+
+    }
+
+    /**
+     * Parses the following portion of user input after the "add" keyword.
+     *
+     * @param inputWithoutAdd The portion of user input containing details of the task to be added.
+     * @param taskList        The TaskList to add the new Task to.
+     * @return The AddCommand which adds the new Task to the TaskList.
+     * @throws InvalidCommandException if the user input is not a valid AddCommand.
+     */
+    protected static AddCommand parseAdd(String inputWithoutAdd, TaskList taskList) throws InvalidCommandException {
```

Comment:
I really like your use of `protected` access modifier to only allow classes in same package which is all that is needed to access such methods in the `Parser` class.